### PR TITLE
h2o for kv cache compression

### PR DESCRIPTION
**Pull request by n1ck-guo**
## Type of Change

feature

## Description

H2O: Heavy-Hitter Oracle for Efficient Generative Inference of Large Language Models
[paper](https://arxiv.org/pdf/2306.14048.pdf)

NTD
- [x] example
- [x] refactor code to same style
- [x] add seq len api
- [x] support for more models, both sim and real
- [x] mean accumulate score function

## Expected Behavior & Potential Risk

None

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

File: intel_extension_for_transformers/transformers/modeling/kv_cahe_compression/h2o_real_drop/modify_llama.py [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+                f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
+                f" and `num_heads`: {self.num_heads})."
+            )
+        self.q_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim, bias=False)
+        self.k_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=False)
+        self.v_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=False)
+        self.o_proj = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias=False)
+        self._init_rope()
+
+        self.kv_cache = H2OKVCache_LayerWise(
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cahe_compression/h2o_real_drop/modify_llama.py:581]: Is there any way to decouple algorithms with the modified kv cache model, for example, keeped_token = token_drop('h2o', kargs, alogirhm_kargs)

File: intel_extension_for_transformers/transformers/modeling/kv_cahe_compression/h2o_real_drop/modify_llama.py [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cahe_compression/h2o_real_drop/modify_llama.py:630]: make sure this file is aligned with the version supported in Gaudi first

File: intel_extension_for_transformers/transformers/modeling/kv_cahe_compression/h2o_sim_drop/modify_gptneox.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        self.register_buffer(
+            "norm_factor",
+            torch.sqrt(torch.tensor(self.head_size, dtype=torch.float32)).to(torch.get_default_dtype()),
+            persistent=False,
+        )
+        self.query_key_value = nn.Linear(config.hidden_size, 3 * config.hidden_size)
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+
+        self.heavy_budget_ratio = config.heavy_ratio
+        self.recent_budget_ratio = config.recent_ratio
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cahe_compression/h2o_sim_drop/modify_gptneox.py:116]: for all the kv pruning methods, let's add one more hyperparameters, seqlen_threshold, if the kept seqlen is less than this threshold, no pruning is conducted.

File: intel_extension_for_transformers/transformers/modeling/kv_cache_compression/h2o.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import OrderedDict
+import importlib
+
+import torch
+from torch import nn
+
+ATTENTION_MAPPING_NAMES = OrderedDict(
+    [
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cache_compression/h2o.py:24]: better use register, then no change in the algorithm code to support new model

Comment (wenhuach21) [reply intel_extension_for_transformers/transformers/modeling/kv_cache_compression/h2o.py:24]: also need to consider the compatible issue for flash attention, sdpa attention

File: intel_extension_for_transformers/transformers/modeling/kv_cache_compression/models/modeling_mistral.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
+                raise ValueError(
+                    f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
+                )
+
+            attn_weights = attn_weights + attention_mask
+
+        # get hh mask
+        if q_len > self.h2o_min_seqlen:
+            mask_bottom = get_hh_mask(self.heavy_ratio, self.recent_ratio, attn_weights)
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cache_compression/models/modeling_mistral.py:147]: better to control this in h2o code

File: examples/huggingface/pytorch/text-generation/h2o/run_lm_eval_harness.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
+import argparse
+import json, tqdm
+import torch
+import copy
+import os
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
+
+import sys
+sys.path.insert(0, '/home/hengguo/code/intel-extension-for-transformers')
```
Comment (wenhuach21) [inline examples/huggingface/pytorch/text-generation/h2o/run_lm_eval_harness.py:9]: don't forget to remove this later

**Comment by n1ck-guo** [comment]
pre-commit.ci autofix

File: intel_extension_for_transformers/transformers/modeling/kv_cache_compression/models/modeling_bloom.py [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+        attention_scores = matmul_result.view(batch_size, self.num_heads, q_length, kv_length)
+
+        input_dtype = attention_scores.dtype
+        # `float16` has a minimum value of -65504.0, whereas `bfloat16` and `float32` have a minimum value of `-3.4e+38`
+        if input_dtype == torch.float16:
+            attention_scores = attention_scores.to(torch.float)
+        attn_weights = torch.masked_fill(attention_scores, attention_mask, torch.finfo(attention_scores.dtype).min)
+
+        # get hh mask
+        from ..h2o import get_hh_mask
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cache_compression/models/modeling_bloom.py:185]: please design an abstract for all the pruning method

File: intel_extension_for_transformers/transformers/modeling/kv_cache_compression/models/modeling_bloom.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+    out = residual + out
+    return out
+
+class H2OBloomForCausalLM(BloomPreTrainedModel):
+    _tied_weights_keys = ["lm_head.weight"]
+
+    def __init__(
+            self,
+            config: BloomConfig,
+            h2o_config: H2OConfig
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cache_compression/models/modeling_bloom.py:248]: remove specific algorithm configs from model file

File: intel_extension_for_transformers/transformers/modeling/kv_cache_compression/prune/__init__.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2024 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
```
Comment (wenhuach21) [inline intel_extension_for_transformers/transformers/modeling/kv_cache_compression/prune/__init__.py:6]: better move these files to other places, they should not be in the modeling folder

**Review by wenhuach21 — APPROVED** [review]

**Comment by PenghuiCheng** [comment]
Could we add a document introducing what h2o is?

**Comment by changwangss** [comment]
format scan will improve by https://github.com/intel/intel-extension-for-transformers/pull/1647.

**Review by changwangss — APPROVED** [review]

**Comment by n1ck-guo** [comment]
> Could we add a document introducing what h2o is?

add in the example/readme

**Review by PenghuiCheng — APPROVED** [review]

**Pull request merged**